### ISSUE TITLE
trial should handle concurrent usage in the same directory gracefully

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2338]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2338|
|Type|enhancement|
|Created|2006-12-30 03:12:59Z|
|Branch|https://github.com/twisted/twisted/tree/concurrent-trial_temp-2338|

Currently a second trial process started in the same directory as a first will destroy the first's _trial_temp directory causing most or all of the remaining tests run by the first process to fail.  trial should handle this case without breaking things in this manner.

<details><summary>Searchable metadata</summary>

```
trac-id__2338 2338
type__enhancement enhancement
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_concurrent_trial_temp_2338 branches/concurrent-trial_temp-2338
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1167448379000000 1167448379000000
changetime__1216843744000000 1216843744000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [24287]) Branching to 'concurrent-trial_temp-2338'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `new`|
|-|-|

okay, maybe that's done.

build results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/concurrent-trial_temp-2338

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Looks good. One minor comment:

```+                testdir = base + '-' + str(counter)```

"%s-%s" would make that more readable.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [24320]) Use interpolation instead of concatenation to create the temp dir name

refs [#2338](https://github.com/twisted/twisted/issues/2338)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [24321]) avoid the repeated stringification

refs [#2338](https://github.com/twisted/twisted/issues/2338)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [24325]) Merge concurrent-trial_temp-2338

Author: exarkun
Reviewer: therve
Fixes: [#2338](https://github.com/twisted/twisted/issues/2338)

Add a lock to the `_trial_temp` directory so that different test
processes don't try to use the same directory at the same time.
If no temporary directory is specified, a different default will
silently be selected if the preferred default is found to be in
use.  If a temporary directory is specified and found to be in
use, an error will be reported.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
